GVFS/GVFS.UnitTests/Maintenance/GitMaintenanceQueueTests.cs

```diff
@@ -0,0 +1,200 @@
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+            queue.EnlistmentRootReady().ShouldBeTrue();
+
+            fileSystem.Paths.Remove(enlistment.EnlistmentRoot);
+            queue.EnlistmentRootReady().ShouldBeFalse();
+
+            fileSystem.Paths.Remove(enlistment.GitObjectsRoot);
+            queue.EnlistmentRootReady().ShouldBeFalse();
+
+            fileSystem.Paths.Add(enlistment.EnlistmentRoot);
+            queue.EnlistmentRootReady().ShouldBeFalse();
+
+            fileSystem.Paths.Add(enlistment.GitObjectsRoot);
+            queue.EnlistmentRootReady().ShouldBeTrue();
+
+            queue.Stop();
+        }
+
+        [TestCase]
+        public void GitMaintenanceQueueHandlesTwoJobs()
+        {
+            ITracer tracer = new MockTracer();
+            GVFSEnlistment enlistment = new MockGVFSEnlistment();
+
+            // We need to have the EnlistmentRoot and GitObjectsRoot available for jobs to run
+            ReadyFileSystem fileSystem = new ReadyFileSystem(new string[]
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            TestGitMaintenanceStep step1 = new TestGitMaintenanceStep(context, gitObjects);
+            TestGitMaintenanceStep step2 = new TestGitMaintenanceStep(context, gitObjects);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+
+            queue.Enqueue(step1);
+            queue.Enqueue(step2);
+
+            Assert.IsTrue(step1.EventTriggered.WaitOne(this.maxWaitTime)
+                && step2.EventTriggered.WaitOne(this.maxWaitTime));
+
+            queue.Stop();
+        }
+
+        [TestCase]
+        public void GitMaintenanceQueueStopSuceedsWhenQueueIsEmpty()
+        {
+            ITracer tracer = new MockTracer();
+            GVFSEnlistment enlistment = new MockGVFSEnlistment();
+
+            // We need to have the EnlistmentRoot and GitObjectsRoot available for jobs to run
+            ReadyFileSystem fileSystem = new ReadyFileSystem(new string[]
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+
+            queue.Stop();

```

Comment:
Is there anything we need to assert here?  Would failure be an exception?

Reply:
Pretty much checking for failure here. Ensuring nothing is `null` or empty during the `Stop()` call.